### DISCUSSION TITLE
Unique Constraints on multiple columns?

### DISCUSSION THREAD

**Discussion by osseonews** [contember/admin#593] [Support]
How would you create a unique constraint on multiple columns with Contember? For example, we store inventory at multiple warehouses. So we  have a join table with WarehouseInventory, with foreign key back to a product_id and warehouse_id. The constraint is that a warehouse_id and product_id are unique, in that you can't have the same product stored twice at the same warehouse. Hope this makes sense.

**Comment by jonasnobile** [comment]
Hi @osseonews, thanks for your question. It's quite easy to define a unique constraint in Contember. You can use the `Unique` decorator to define unique constraint on multiple columns ([see](https://docs.contember.com/reference/engine/schema/columns#unique-fields)).

Here is simplified code how to solve your example:
```ts
import { SchemaDefinition as def } from '@contember/schema-definition'

@def.Unique('product', 'warehouse')
export class WarehouseInventory {
  product = def.manyHasOne(Product, 'inventory')
  warehouse = def.manyHasOne(Warehouse, 'inventory')
}

**Comment by osseonews** [comment]
Thanks for the quick response! Instead of opening up a new discussion, I had one other quick question. Where are the images stored in Contember cloud? When we create an Image field, it automatically creates an upload and we can upload images. But where are these images and how to access them outside of cloud? Is there an Image API of some sort? Can we export the images and host elsewhere?

**Reply by jonasnobile** [reply]
All images (and all other objects, e.g. files) are stored in S3. Contember stores the public url in the image entity, so you can use it to access the image anywhere else. If you want your images to be private, you set this up and access the images later with the `generateReadUrl' mutation ([see](https://docs.contember.com/reference/engine/content/s3)).

We do not currently provide any kind of image API. We do have a custom S3 feature in the backlog, so if there is demand for it, we will be happy to implement it.

**Reply by osseonews** [reply]
Thanks. So I guess, we can upload the images with Contember, then with an Action call a webhook which would then get the image and upload it to another server? Is that a possible workaround? BTW, your AI is simply incredible and a huge timesaver. I had looked at Contember around a year ago, but it was a bit too complex to set up. Now with the AI and Cloud, this is simply an incredible system. Congrats!

**Reply by matej21** [reply]
If you wish, you can entirely replace the "uploader" in your admin interface and upload it directly to your server.
Though this isn't currently documented, you can implement the [FileUploader interface](https://github.com/contember/interface/blob/main/packages/client/src/content/upload/FileUploader.ts). For guidance, you can check the [default S3 uploader](https://github.com/contember/interface/blob/main/packages/client/src/content/upload/S3FileUploader.ts) and use it as a reference.
Then you just set the `uploader` prop of upload components.

**Reply by honzasladek** [reply]
Btw quick question @osseonews: what's your use case to upload images elsewhere? (Just curious here.)

**Reply by osseonews** [reply]
We use Cloudflare for images. It's just much better than S3 for various reasons. So we need to upload the images to cloudflare and then we get the Cloudflare Image id and can call it in our app. https://www.cloudflare.com/developer-platform/cloudflare-images/

**Reply by honzasladek** [reply]
@osseonews Thanks! We use Cloudflare Images as well but keep the images in S3 - works like a charm. Here's a very simple Cloudflare worker that will do the job (whitelists the `data.eu.cntmbr.com` where Contember stores the data.). It just a simple example (though works fine) for inspiration.

```
addEventListener('fetch', event => {
  event.respondWith(handleRequest(event.request))
})


async function handleRequest(request) {
  const whiteListOrigins = [
    'data.eu.cntmbr.com',
    'www.whatever-domain-you-want.com',
  ]
  const requestURL = new URL(request.url)
  const pathname = requestURL.pathname.split('/')

  // Check if source origin is allowed
  if(!whiteListOrigins.includes(pathname[1])) {
    return new Response("Your origin is not in allowed list.", { status: 403 })
  }

  const accept = request.headers.get('accept')
  let defaultFormat
  if (/image\/avif/.test(accept)) {
    defaultFormat = 'avif'
  } else if (/image\/webp/.test(accept)) {
    defaultFormat = 'webp'
  }

  // Get resize options from searchParams
  const width = requestURL.searchParams.get('width')
  const height = requestURL.searchParams.get('height')
  const format = requestURL.searchParams.get('format') ?? defaultFormat
  const quality = requestURL.searchParams.get('quality')
  const fit = requestURL.searchParams.get('fit')

  // Remove first /
  pathname.shift()

  return fetch('https://' + pathname.join('/'),
    {
      cf: {
        image: {
          width,
          height,
          format,
          quality,
          fit,
        }
      },
    }
  )
}

```
